src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java

```diff
@@ -49,33 +44,80 @@ public String description()
 			return Collections.emptyList();
 		}
 
-		if (log.isDebugEnabled()) {
-			log.debug("getInstances: service = {}", service.toString());
+		return getInstancesFromService(service);
+	}
+
+	@Override
+	public List<ServiceInstance> selectInstances(Map<String, String> match)
+	{
+		return selectInstances(match, Collections.emptyMap());
+	}
+
+	@Override
+	public List<ServiceInstance> selectInstances(Map<String, String> match, Map<String, String> doNotMatch)
+	{
+		ServiceList serviceList;
+		try {
+			serviceList = kubeClient.services().withLabels(match).withoutLabels(doNotMatch).list();
+		} catch(KubernetesClientException e) {
+			log.warn("getInstancesMatchingMetadata: failed to retrieve matching services: API call failed: {}",
+				e.getMessage());
+			return Collections.emptyList();
 		}
 
-		if (service.getSpec().getPorts().isEmpty()) {
-			log.error("getInstances: service '{}' has no ports", serviceId);
+		if (serviceList == null) {
+			log.warn("getInstancesMatchingMetadata: service list is null");
 			return Collections.emptyList();
 		}
 
-		List<ServicePort> servicePorts = service.getSpec().getPorts();
-		ServicePort svcPort = servicePorts.get(0); // By default, use the first port available
+		return getInstancesFromServiceList(serviceList.getItems());
+	}
+
+	private List<ServiceInstance> getInstancesFromService(Service service)
+	{
+		List<Service> serviceList = new ArrayList<>();
+		serviceList.add(service);
+		return getInstancesFromServiceList(serviceList);
+	}
 
-		// But if a port named "http" is available, use it instead
-		Optional<ServicePort> httpPort = servicePorts.stream()
-			.filter(s -> s.getName() != null && s.getName().equals("http")).findFirst();
-		if (httpPort.isPresent()) {
-			svcPort = httpPort.get();
+	private List<ServiceInstance> getInstancesFromServiceList(List<Service> services)
+	{
+		if (log.isDebugEnabled()) {
+			log.debug("services = {}", services.toString());
 		}
 
 		List<ServiceInstance> serviceInstances = new ArrayList<>();
-		serviceInstances.add(new DefaultServiceInstance(
-			service.getMetadata().getName(),
-			service.getSpec().getClusterIP(),
-			svcPort.getPort(),
-			false,
-			service.getMetadata().getLabels()
-		));
+		for(Service service : services)
+		{
+			if (service.getSpec() == null) {
+				log.error("skipping service with no spec");
+				continue;
+			}
+
+			String serviceName = service.getMetadata().getName();
+			if (service.getSpec().getPorts().isEmpty()) {
+				log.error("service '{}' has no ports", serviceName);
+				continue;
+			}
+
+			List<ServicePort> servicePorts = service.getSpec().getPorts();
+			ServicePort svcPort = servicePorts.get(0); // By default, use the first port available
+
+			// But if a port named "http" is available, use it instead
+			Optional<ServicePort> httpPort = servicePorts.stream()
```

Comment:
I find this whole construct a little bit clumsy, I would expect steps like this:

1. if service has only one port, use that one no matter what the name is
2. if service has multiple ports, look for the one with http name
3. if port with http name is found, use that one
4. if port with http name is not found, issue warning to logs (not sure about the level here, either INFO or WARN ...) and use the first one